### ISSUE TITLE
Json Encoders are ignored in nested structures

### ISSUE THREAD

**Issue by mvanderlee** [pydantic/pydantic#2277]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

When nesting a model with customer encoders, they are ignored by the parent.

```python
from datetime import datetime, timedelta 
from pydantic import BaseModel 
from pydantic.json import timedelta_isoformat 
 
 
class WithCustomEncoders(BaseModel): 
    dt: datetime 
    diff: timedelta 
 
    class Config: 
        json_encoders = { 
            datetime: lambda v: v.timestamp(), 
            timedelta: timedelta_isoformat, 
        } 
 
class ParentWithoutEncoders(BaseModel): 
    child: WithCustomEncoders
 
m = WithCustomEncoders(dt=datetime(2032, 6, 1), diff=timedelta(hours=100)) 
print(m.json())

p = ParentWithoutEncoders(child=m)
print(p.json())


#> {"dt": 1969671600.0, "diff": "P4DT4H0M0.000000S"}
#> {"child": {"dt": "2032-06-01T00:00:00", "diff": 360000.0}}
```

Expected output:
```python
#> {"dt": 1969671600.0, "diff": "P4DT4H0M0.000000S"}
#> {"child": {"dt": 1969671600.0, "diff": "P4DT4H0M0.000000S"}}
```

**Comment by PrettyWood** [comment]
Hello @mvanderlee 
The config depends on the model not the type of the field, even if this one is also a model.
So you'll need to duplicate your config or add it in your own custom BaseModel.

**Comment by mvanderlee** [comment]
Duplicating the config is a non-starter for us. 
So you're saying that there is currently no way to ensure a Model will always be serialized a certain way? Other than manually defining a json encoder and always specifying it on json.dump. Which again, would be a non-started for us. 

I need to allow Models to be 2-way serializable.

**Comment by PrettyWood** [comment]
Sure thing!
```python
from datetime import datetime, timedelta
from pydantic import BaseModel as PydanticBaseModel
from pydantic.json import timedelta_isoformat


class BaseModel(PydanticBaseModel):
    """
    All the instances of BaseModel should serialize
    those types the same way
    """
    class Config:
        json_encoders = {
            datetime: lambda v: v.timestamp(),
            timedelta: timedelta_isoformat,
        }


class WithCustomEncoders(BaseModel):
    dt: datetime
    diff: timedelta


class ParentWithoutEncoders(BaseModel):
    child: WithCustomEncoders


m = WithCustomEncoders(dt=datetime(2032, 6, 1), diff=timedelta(hours=100))
print(m.json())
# {"dt": 1969653600.0, "diff": "P4DT4H0M0.000000S"}

p = ParentWithoutEncoders(child=m)
print(p.json())
# {"child": {"dt": 1969653600.0, "diff": "P4DT4H0M0.000000S"}}
```
The magic is possible because `Config` of a subclass inherits from the `Config` of its parent!

**Comment by shyamd** [comment]
I agree with @mvanderlee that it's not clear at first that the parent model that is converting to JSON establishes all the encoders and the nested models are converted using those. This appears to be because the whole object is converted into a dict and then the JSON encoder of the parent model is applied to the dictionary. An alternative might be to recursively call the JSON encoder on any pydantic BaseModel children and incorporate along the way.

@mvanderlee if you're interested in a PR, you might look at modifying https://github.com/samuelcolvin/pydantic/blob/13a5c7d676167b415080de5e6e6a74bea095b239/pydantic/main.py#L474-L509

This is likely to be much slower than the single call to JSON encoder, so I'd recommend making it an option or even a new method?

**Comment by goroderickgo** [comment]
Agreed with the other folks here that there may be use-cases where you might want to serialize different components of the same type differently. Based on @PrettyWood's example, what I came up with was just to create subclasses of the nested object (maybe this is overkill, would appreciate any suggestions!). Using a similar example:

```
class DatetimeA(datetime.datetime):
    """Datetime that will be encoded as a string in pydantic Config."""
    def __new__(cls, *args, **kwargs):
        return datetime.datetime.__new__(cls, *args, **kwargs)

class DatetimeB(datetime.datetime):
    """Datetime that will be encoded as a timestamp in pydantic Config."""
    def __new__(cls, *args, **kwargs):
        return datetime.datetime.__new__(cls, *args, **kwargs)

class MyBaseModel(BaseModel):
    class Config:
        json_encoders = {
            DatetimeA: lambda v: str(v),
            DatetimeB: lambda v: v.timestamp()
    }

class TestModel(MyBaseModel):
    datetime_a: DatetimeA
    datetime_b: DatetimeB

class ParentModel(MyBaseModel):
    test_model: TestModel

test_model = TestModel(
    datetime_a=DatetimeA(2020, 1, 1),
    datetime_b=DatetimeB(2015, 1, 1)
)

parent_model = ParentModel(
    test_model=test_model
)

print(test_model.json())
print(parent_model.json())
```
Output:
```
# '{"datetime_a": "2020-01-01 00:00:00", "datetime_b": 1420088400.0}'
# '{"test_model": {"datetime_a": "2020-01-01 00:00:00", "datetime_b": 1420088400.0}}'
```

Perhaps I haven't been looking at the right parts of the documentation, but I didn't really understand the `Config` inheritance until I stumbled upon this issue and saw [@PrettyWood's explanation](https://github.com/samuelcolvin/pydantic/issues/2277#issuecomment-764010272). Would it be possible to add some of this discussion to the documentation if it's not already there?

**Comment by lilyminium** [comment]
I opened a PR at #3941 addressing this with a keyword argument (`use_nested_encoders`). It's a fairly simple band-aid onto the issue and probably the top flaw is that it does not work when `models_as_dict=False`, because that just passes raw data into `json.dumps` with a global encoder. Would love to get feedback, or continue discussing here!

**Comment by samuelcolvin** [comment]
I can see the argument for this, but it will have to wait until v2.

**Comment by lilyminium** [comment]
Thanks -- can you let me know how that affects #3941? Should it (continue being) on hold?

**Comment by samuelcolvin** [comment]
I've changed my mind, @lilyminium you PR looks great, needs some fixes, but I'd be happy to accept it for v1.10.

**Comment by ryzhakar** [comment]
Just to clarify for a passing-by reader: this doesn't seem to be solved. The PR above was initially accepted and then reverted. Now the problems waits till the V2 release.

**Comment by samuelcolvin** [comment]
Correct, sorry for confusion. 

See #4456.

**Comment by samuelcolvin** [comment]
V2 should be released some time in Q1 2023. I had hoped to have it out this year, but I haven't had as much help as I'd hoped for.

**Comment by aguckenber-chwy** [comment]
>I've changed my mind, @lilyminium you PR looks great, needs some fixes, but I'd be happy to accept it for v1.10.

Anyway this part can be part of V1?

**Comment by samuelcolvin** [comment]
No, v1 develop has stopped except for bug fixes.

**Comment by aguckenber-chwy** [comment]
So much for your comment above, I guess you had a change of heart. Will wait for V2 and fork it to fix it.

**Comment by samuelcolvin** [comment]
Well I think it was more a omission, the pr wasn't completed in time for v1.10, it's too big to be included in a patch release and there won't be a v1.11. Thus by default the change has to wait for V2.

Remember pydantic isn't (yet) properly funded and it's developed and maintained by volunteers.

Thanks for your understanding. I'm working flat out on V2, no one wants it released more than me.

**Comment by aguckenber-chwy** [comment]
> Well I think it was more a omission, the pr wasn't completed in time for v1.10, it's too big to be included in a patch release and there won't be a v1.11. Thus by default the change has to wait for V2.
> 
> Remember pydantic isn't (yet) properly funded and it's developed and maintained by volunteers.
> 
> Thanks for your understanding. I'm working flat out on V2, no one wants it released more than me.

I really do appreciate the work by the way, and the library. I was just disappointed to google my issue, find a branch that was merged 1 month prior to v1.10 that fixed it, but wasn't released in 1.10, wasting 6 hours trying to find a workaround because Pydantic doesn't listen to nested encoders and dealing with circular references has been a pain, but I really appreciate the work you do! Keep it up! Sorry about the tone. Just upset that it didn't make it in V.10.

**Comment by dmontagu** [comment]
It seems to me that the specific issues brought up in this issue can mostly be resolved now through the use of `@model_serializer` and `@field_serializer`. We have actually removed the `Config.json_encoders` thing in v2 in favor of this approach, but if you run into problems caused by the removal of `json_encoders`, please open a new issue about it and we'll be happy to attempt to resolve your issues and/or come up with a way to further ease migration.
